### PR TITLE
feature: Step by step sync

### PR DESCRIPTION
**Pull request by calmbeing**
Update the way to sync step by step, which can support both step-by-step sync and sync upper-bound set

### CODE REVIEW

File: eth/stagedsync/stage_headers.go [hunk 1]
```diff
@@ -110,6 +110,11 @@ func SpawnStageHeaders(
 		}
 		defer tx.Rollback()
 	}
+	if initialCycle {
```
Comment (setunapo) [inline eth/stagedsync/stage_headers.go:113]: initialCycle == true, only for the first time?

File: turbo/stages/headerdownload/header_algos.go [hunk 2]
```diff
@@ -608,6 +608,9 @@ func (hd *HeaderDownload) InsertHeader(hf FeedHeaderFunc, terminalTotalDifficult
 			blocksToTTD = x.Uint64()
 		}
 	}
+	if hd.stageSyncStep > 0 && hd.highestInDb%hd.stageSyncStep == 0 {
```
Comment (setunapo) [inline turbo/stages/headerdownload/header_algos.go:611]: why put it here, could be the same place as stageSyncUpperBound

Comment (calmbeing) [reply turbo/stages/headerdownload/header_algos.go:611]: If put as the same place above, then it was check before insert, then the blockHeight whose mod stageSyncStep ==0 will never be inserted into db, so put here, insert before check.

File: eth/stagedsync/stage_headers.go (hunk 1 above)
Comment (calmbeing) [reply eth/stagedsync/stage_headers.go:113]: Yes, only for the first time, then will always be false afterwards

Comment (alexqrid) [reply eth/stagedsync/stage_headers.go:113]: Look through [the issue](https://github.com/ledgerwatch/erigon/issues/7339) please and [PR](https://github.com/ledgerwatch/erigon/pull/7532) this is crucial for lagging networks like BSC. You can find more at erigon's discord - https://discord.com/channels/687972960811745322/983710293156851763/1108134331882410045

Comment (setunapo) [reply eth/stagedsync/stage_headers.go:113]: @alexqrid pls create an issue to track your issue. Your reply is irrelevant to this PR.

**Review by setunapo — APPROVED** [review]

**Pull request merged**